src/batch.rs

```diff
@@ -0,0 +1,297 @@
+//! Performs batch RedJubjub signature verification.
+//!
+//! Batch verification asks whether *all* signatures in some set are valid,
+//! rather than asking whether *each* of them is valid. This allows sharing
+//! computations among all signature verifications, performing less work overall
+//! at the cost of higher latency (the entire batch must complete), complexity of
+//! caller code (which must assemble a batch of signatures across work-items),
+//! and loss of the ability to easily pinpoint failing signatures.
+//!
+
+use std::convert::TryFrom;
+
+use jubjub::*;
+use rand_core::{CryptoRng, RngCore};
+
+use crate::{private::Sealed, scalar_mul::VartimeMultiscalarMul, *};
+
+// Shim to generate a random 128bit value in a [u64; 4], without
+// importing `rand`.
+fn gen_128_bits<R: RngCore + CryptoRng>(mut rng: R) -> [u64; 4] {
+    let mut bytes = [0u64; 4];
+    bytes[0] = rng.next_u64();
+    bytes[1] = rng.next_u64();
+    bytes
+}
+
+enum Inner {
+    SpendAuth {
+        vk_bytes: VerificationKeyBytes<SpendAuth>,
+        sig: Signature<SpendAuth>,
+        c: Scalar,
+    },
+    Binding {
+        vk_bytes: VerificationKeyBytes<Binding>,
+        sig: Signature<Binding>,
+        c: Scalar,
+    },
+}
+
+impl
+    From<(
+        VerificationKeyBytes<SpendAuth>,
+        Signature<SpendAuth>,
+        Scalar,
+    )> for Inner
+{
+    fn from(
+        tup: (
+            VerificationKeyBytes<SpendAuth>,
+            Signature<SpendAuth>,
+            Scalar,
+        ),
+    ) -> Self {
+        let (vk_bytes, sig, c) = tup;
```

Comment:
TIOLI: You can do the destructuring in the fn signature
```suggestion
        (vk_bytes, sig, c): (
            VerificationKeyBytes<SpendAuth>,
            Signature<SpendAuth>,
            Scalar,
        ),
    ) -> Self {
```